### PR TITLE
Fix chat rooms, add hooks and eventmacro conditions to them

### PR DESCRIPTION
**Pull request by Henrybk**

### CODE REVIEW

File: src/Network/Receive/ServerType0.pm [hunk 1]
```diff
@@ -145,7 +145,7 @@ sub new {
 		'00DB' => ['chat_users'],
 		'00DC' => ['chat_user_join', 'v Z24', [qw(num_users user)]],
 		'00DD' => ['chat_user_leave', 'v Z24 C', [qw(num_users user flag)]],
-		'00DF' => ['chat_modified', 'x2 a4 a4 v2 C a*', [qw(ownerID ID limit num_users public title)]],
+		'00DF' => ['chat_modified', 'v a4 a4 v2 C', [qw(len ownerID ID limit num_users public)]], # -1
```
Comment (alisonrag) [inline src/Network/Receive/ServerType0.pm]: this packet does not send chat tittle anymore?

Comment (Henrybk) [inline src/Network/Receive/ServerType0.pm]: It is defined in the actual sub in Receive.pm

File: plugins/eventMacro/eventMacro/Condition/ChatRoomNear.pm [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+		$self->{fulfilled_chatID} = $ID;
+		last;
+	}
+}
+
+sub get_new_variable_list {
+	my ($self) = @_;
+	my $new_variables;
+	
+	$new_variables->{".".$self->{name}."Last"."ID"} = $self->{fulfilled_chatID};
```
Comment (allanon) [inline plugins/eventMacro/eventMacro/Condition/ChatRoomNear.pm]: Why use the concatenation operator? `$new_variables->{".$self->{name}LastID"}`

Also, if this is a thing, and it seems like it is, how about making a helper to make setting these variables easier?

```
# Helper to assign "$.Last*" variables easier.
sub _last_variables {
    my ($self, $vars) = @_;

    my $new_variables = {};
    $new_variables->{".$self->{name}Last$_"} = $vars{$_} foreach keys %$vars;
    $new_variables;
}

sub get_new_variable_list {
    my ($self) = @_;
    my ($index) = grep { defined $chatRoomsID[$_] eq $self->{fulfilled_chatID} } 0..$#chatRoomsID;
    my $new_variables = $self->_last_variables({
        ID => $self->{fulfilled_chatID},
        OwnerID => $chatRooms{$self->{fulfilled_chatID}}{ownerID},
        OwnerName => Actor::get($chatRooms{$self->{fulfilled_chatID}}{ownerID})->name,
        Title => $chatRooms{$self->{fulfilled_chatID}}{title},
        BinID => $index,
    });
}
```

File: plugins/eventMacro/eventMacro/Condition/InChatRoom.pm [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+use Globals;
+
+sub _hooks {
+	['packet_mapChange','chat_created','chat_leave','chat_joined'];
+}
+
+sub _parse_syntax {
+	my ( $self, $condition_code ) = @_;
+	
+	$self->{wanted_state} = undef;
```
Comment (allanon) [inline plugins/eventMacro/eventMacro/Condition/InChatRoom.pm]: The default value of `wanted_state` is `undef`. And the value doesn't matter if this function returns zero anyway, I think?

How about:
```
if ($condition_code !~ /^[01]$/) {
    $self->{error} = "Value '$condition_code' should be '0' or '1'";
    return 0;
}

$self->{wanted_state} = $condition_code;

return 1;
```

File: src/Network/Receive.pm [hunk 4]
```diff
@@ -2896,24 +2900,41 @@ sub chat_join_result {
...
+	my $title_len = ($args->{len} - $header_len);
+	
+	my $title = substr($msg, $header_len, $title_len);
 
-	my $title = bytesToString($args->{title});
-
-	my ($ownerID, $ID, $limit, $public, $num_users) = @{$args}{qw(ownerID ID limit public num_users)};
-
+	my ($ownerID, $chat_ID, $limit, $public, $num_users) = @{$args}{qw(ownerID ID limit public num_users)};
+
```
Comment (allanon) [inline src/Network/Receive.pm]: Unnecessary whitespace.

File: src/Network/Receive.pm [hunk 5]
```diff
@@ -2896,24 +2900,41 @@ sub chat_join_result {
...
 	my ($self, $args) = @_;
+	
+	my $msg = $args->{RAW_MSG};
+	my $msg_size = $args->{RAW_MSG_SIZE};
+	my $header_pack = 'v a4 a4 v2 C';
+	my $header_len = ((length pack $header_pack) + 2);
+	
+	my $title_len = ($args->{len} - $header_len);
+	
+	my $title = substr($msg, $header_len, $title_len);
```
Comment (allanon) [inline src/Network/Receive.pm]: `$title_len` is unnecessary. `substr $str, $offset` returns the rest of the string, so this is just:

```
my $title = bytesToString(substr $msg, $header_len);
```

I believe `bytesToString` is necessary for utf8 support.

File: src/Network/Receive/ServerType0.pm (hunk 1 above)
Comment (allanon) [inline src/Network/Receive/ServerType0.pm]: It's better practice to put all of the available fields into the unpack string here, if possible. In addition to being more clear and concise, it also helps people implementing their own packet handlers (via `packet/*` and `packet_pre/*` hooks). Why remove `title`?

File: plugins/eventMacro/eventMacro/Condition/ChatRoomNear.pm (hunk 2 above)
Comment (Henrybk) [inline plugins/eventMacro/eventMacro/Condition/ChatRoomNear.pm]: This ia good idea for the future
